src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -1355,3 +1373,36 @@ InsertSelectResultIdPrefix(uint64 planId)
 
 	return resultIdPrefix->data;
 }
+
+
+/*
+ * GetInsertSelectMethod returns the preferred INSERT INTO ... SELECT method
+ * based on its select query.
+ */
+static InsertSelectMethod
+GetInsertSelectMethod(Query *selectQuery, Oid targetRelationId, ParamListInfo boundParams)
+{
+	Query *selectQueryCopy = copyObject(selectQuery);
+
+	/*
+	 * Query will be replanned in insert_select_executor to plan correctly
+	 * for prepared statements. So turn off logging here to avoid repeated
+	 * log messages. We use SET LOCAL here so the change is reverted on ERROR.
+	 */
+	int savedClientMinMessages = client_min_messages;
+	set_config_option("client_min_messages", "ERROR",
+					  PGC_USERSET, PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
+
+	int cursorOptions = CURSOR_OPT_PARALLEL_OK;
+	PlannedStmt *selectPlan = pg_plan_query(selectQueryCopy, cursorOptions,
```

Comment:
I think the reason we put pg_plan_query in insert_select_executor.c rather than insert_select_planner.c is that we also needed insertSelectQuery in the executor and we had insertSelectQuery, there wasn't really a reason to call pg_plan_query in the planner.

Given how much redundant computation work we are doing, would it make sense to store both insertSelectQuery and selectPlanForInsertSelect in the DistributedPlan?

Reply:
done